swaybar/tray/menu.c

```diff
@@ -0,0 +1,975 @@
+			sd_bus_message_enter_container(msg, 'v', "(ia{sv}av)");
+		}
+	}
+
+	struct swaybar_popup *popup = sni->tray->popup;
+	if (ret < 0) {
+		sway_log(SWAY_ERROR, "%s%s failed to read menu layout: %s",
+				sni->service, sni->menu_path, strerror(-ret));
+		destroy_menu(sni->menu);
+		sni->menu = NULL;
+	} else if (popup->sni == sni) {
+		if (popup->popup_surface) {
+			close_popup(popup); // TODO enhancement: redraw instead of closing
+		} else {
+			open_popup_id(sni, 0);
+		}
+	}
+	return ret;
+}
+
+static void update_menu(struct swaybar_sni *sni, int id) {
+	struct swaybar_sni_slot *slot = calloc(1, sizeof(struct swaybar_sni_slot));
+	slot->sni = sni;
+
+	int ret = sd_bus_call_method_async(sni->tray->bus, &slot->slot,
+			sni->service, sni->menu_path, menu_interface, "GetLayout",
+			get_layout_callback, slot, "iias", id, -1, NULL);
+	if (ret >= 0) {
+		wl_list_insert(&sni->slots, &slot->link);
+	} else {
+		sway_log(SWAY_ERROR, "%s%s failed to get layout: %s",
+				sni->service, sni->menu_path, strerror(-ret));
+		free(slot);
+	}
+}
+
+static int handle_layout_updated(sd_bus_message *msg, void *data,
+		sd_bus_error *error) {
+	struct swaybar_sni *sni = data;
+	sway_log(SWAY_DEBUG, "%s%s layout updated", sni->service, sni->menu_path);
+
+	int id;
+	sd_bus_message_read(msg, "ui", NULL, &id);
+	update_menu(sni, id);
+	return 0;
+}
+
+static int handle_items_properties_updated(sd_bus_message *msg, void *data,
+		sd_bus_error *error) {
+	struct swaybar_sni *sni = data;
+	sway_log(SWAY_DEBUG, "%s%s item properties updated", sni->service, sni->menu_path);
+
+	// update properties
+	sd_bus_message_enter_container(msg, 'a', "(ia{sv})");
+	while (!sd_bus_message_at_end(msg, 0)) {
+		sd_bus_message_enter_container(msg, 'r', "ia{sv}");
+		int id;
+		sd_bus_message_read_basic(msg, 'i', &id);
+		update_item_properties(*menu_find_item(&sni->menu, id), msg);
+	}
+
+	// removed properties
+	sd_bus_message_enter_container(msg, 'a', "(ias)");
+	while (!sd_bus_message_at_end(msg, 0)) {
+		sd_bus_message_enter_container(msg, 'r', "ia{sv}");
+		int id;
+		sd_bus_message_read_basic(msg, 'i', &id);
+		struct swaybar_menu_item *item = *menu_find_item(&sni->menu, id);
+
+		char **keys;
+		sd_bus_message_read_strv(msg, &keys);
+		if (keys) {
+			for (char **key = keys; *key; ++key) {
+				if (strcmp(*key, "type") == 0) {
+					item->is_separator = false;
+				} else if (strcmp(*key, "label") == 0) {
+					free(item->label);
+					item->label = NULL;
+				} else if (strcmp(*key, "enabled") == 0) {
+					item->enabled = true;
+				} else if (strcmp(*key, "visible") == 0) {
+					item->visible = true;
+				} else if (strcmp(*key, "children-display") == 0) {
+					for (int i = 0; i < item->children->length; ++i) {
+						destroy_menu(item->children->items[i]);
+					}
+					list_free(item->children);
+				}
+			}
+		}
+	}
+
+	struct swaybar_popup *popup = sni->tray->popup;
+	if (popup->sni == sni) {
+		close_popup(popup); // TODO enhancement: redraw instead of closing
```

Comment:
Clicking an item on the first tray menu interaction with Zoom's icon consistently crashes.

The issue is due to `ItemsPropertiesUpdated` signalling during the first interaction; Zoom updates the "Choose a Language" submenu to indicate the chosen language.

What happens is that `close_popup` is called before the `AboutToShow` callback triggers, and so the popup remains open and clicking on it causes `SIGSEGV`, as `close_popup` has set `sni = NULL`.